depends: Explicitly download and vendor Rust dependencies

str4d opened the pull request:
Closes #2231.

str4d commented:
Alternative to #3094.

ebfull commented:
Add --frozen to cargo build

ebfull commented:
@zkbot try

str4d commented:
Failed due to an odd ccache error during the Debian 9 build that I haven't seen before.

ebfull commented on depends/funcs.mk:
```diff
@@ -30,6 +30,23 @@ define fetch_file
+
+define configure_crate_registry
+mkdir .cargo && \
+echo "[source.vendored-sources]\ndirectory = \"$(host_prefix)/$(CRATE_REGISTRY)\"\n\n[source.crates-io]\nreplace-with = \"vendored-sources\"" > .cargo/config
```
These newlines aren't going to work right depending on the platform. (In the homu try, you can see the mac build failed because these weren't escaped.)

ebfull replied:
I think it would be better to have a template file that gets sed'd, allowing us to easily change other configurations or target features.

str4d commented:
> These newlines aren't going to work right depending on the platform. (In the homu try, you can see the mac build failed because these weren't escaped.)

The MacOS build actually failed because of issues in the command that generates the `.cargo-checksum.json` file. It seems that the intermediate `sed` that converts `sha256sum` output `H(A)  A` to `"A":"H(A)"` is not working. I'll dig into that.

str4d commented:
@ebfull addressed your comments.

@zkbot try

str4d commented:
I can squash the last two commits into the first one if desired.

daira commented on depends/funcs.mk:
```diff
@@ -30,6 +30,18 @@ define fetch_file
 
+define generate_crate_checksum
+rm .gitignore && \
+echo "{\"files\":{`find . -type f | grep -v "$($(1)_file_name)" | grep -v .stamp_ | sed 's|^./||' | sort | xargs $(build_SHA256SUM) | awk -v OFS='":"' '{print $$$$2, $$$$1}' | sed 's|^|"|' | sed 's|$|"|' | tr '\n' ',' | sed 's|,$$$$||'`},\"package\":\"$($(1)_sha256_hash)\"}" > .cargo-checksum.json
```
Oww my head. Sorry, the interaction of quoting levels and of shell vs make expansion here is just impossible to review. NACK until it is simplified somehow (maybe an intermediate variable for the result of the backticked expression?)

daira replied:
It doesn't help that it's all on one line.

daira replied:
Or maybe a separate shell script?

str4d commented:
@daira moved the `.cargo-checksum.json` generation into a separate script.

str4d commented:
Hmm, after addressing comment from @ebfull, MacOS still breaks:

```
echo Preprocessing librustzcash...
Preprocessing librustzcash...
mkdir -p /Users/zcbbworker/macos-0/macos10-12/build/depends/work/build/x86_64-apple-darwin16.7.0/librustzcash/0.1-b76c4e1aec6 /Users/zcbbworker/macos-0/macos10-12/build/depends/work/build/x86_64-apple-darwin16.7.0/librustzcash/0.1-b76c4e1aec6/.patches-b76c4e1aec6
cd /Users/zcbbworker/macos-0/macos10-12/build/depends/patches/librustzcash; cp cargo.config /Users/zcbbworker/macos-0/macos10-12/build/depends/work/build/x86_64-apple-darwin16.7.0/librustzcash/0.1-b76c4e1aec6/.patches-b76c4e1aec6 ;
cd /Users/zcbbworker/macos-0/macos10-12/build/depends/work/build/x86_64-apple-darwin16.7.0/librustzcash/0.1-b76c4e1aec6;   mkdir .cargo && cp /Users/zcbbworker/macos-0/macos10-12/build/depends/work/build/x86_64-apple-darwin16.7.0/librustzcash/0.1-b76c4e1aec6/.patches-b76c4e1aec6/cargo.config .cargo/config && sed -i 's|CRATE_REGISTRY|/Users/zcbbworker/macos-0/macos10-12/build/depends/x86_64-apple-darwin16.7.0/vendored-sources|' .cargo/config
sed: 1: ".cargo/config": invalid command code .
```

I'll try streaming the patch instead.

str4d commented:
@zkbot try

str4d commented:
That worked. All builders successfully built the dependencies.

daira commented on depends/cargo-checksum.sh:
```diff
@@ -0,0 +1,13 @@
+awk -v OFS='":"' '{print $2, $1}' |	#   'H(A) A'  ->  'A":"H(A)'
+sed 's|^|"|' |				#  'A":"H(A)' -> '"A":"H(A)'
+sed 's|$|"|' |				# '"A":"H(A)' -> '"A":"H(A)"'
+tr '\n' ',' |		# Concatenate lines with commas
```
I think backslash-escaping within backticks is nonportable (POSIX allows the backslash sequence to be expanded early).

http://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_xcu_chap02.html#tag_23_02_06_03

daira replied:
See also https://stackoverflow.com/questions/43838007/posix-shell-escaping-line-continuations-in-backquote-command-substitutions

str4d replied:
I used backticks originally because of the Makefile dollar-symbol issue. I'll change it now.

str4d commented:
@zkbot try

(I expect this will fail due to PyFlakes warnings in master, as we now requiring the code to be PyFlakes-clean, but dependencies should build on all builders.)

str4d commented:
Two builders failed due to transient errors, but the other two failed due to PyFlakes warnings (yay!)

mdr0id commented:
@str4d Hit an error that I saw in an earlier PR: https://github.com/zcash/zcash/pull/3117/commits
> Makefile:154 recipe for target 'download-win' failed

After applying 3117 fix, I am able to test the above without failure:
> make -C depends download

ebfull commented:
ACK from me!

str4d commented:
Squashed commits at @daira's request.

str4d commented:
@zkbot r+

daira commented on depends/cargo-checksum.sh:
```diff
@@ -0,0 +1,13 @@
+find . -type f |  	# Get list of file paths
+grep -v $1 |		# Exclude Makefile hashes
+grep -v .stamp_ |	# Exclude Makefile stamps
+sed 's|^./||' |		# Remove leading ./
```
Spotted a bug post-merge: this is supposed to match only leading "./", but actually matches any leading character followed by "/" as the second character.

```
$ echo 'X/foo' |sed 's|^./||'  # incorrect
foo
$ echo 'X/foo' |sed 's|^[.]/||'  # correct
X/foo
$ echo './foo' |sed 's|^[.]/||'  # correct
foo
```

daira replied:
Similarly, the previous line should be ``grep -v '[.]stamp_' |``.

str4d replied:
Opened #3131.
